src/socket.rs

```diff
@@ -84,13 +84,101 @@ impl ::IntoInner for Socket {
     fn into_inner(self) -> sys::Socket { self.inner }
 }
 
-fn addr2raw(addr: &SocketAddr) -> (*const c::sockaddr, c::socklen_t) {
+/// A type with the same memory layout as `c::sockaddr`. Used in converting Rust level
+/// SocketAddr* types into their system representation. The benefit of this specific
+/// type over using `c::sockaddr_storage` is that this type is exactly as large as it
+/// needs to be and not a lot larger. And it can be initialized cleaner from Rust.
+#[repr(C)]
+pub(crate) union SocketAddrCRepr {
+    v4: c::sockaddr_in,
+    v6: c::sockaddr_in6,
+}
+
+impl SocketAddrCRepr {
+    pub(crate) fn as_ptr(&self) -> *const c::sockaddr {
+        self as *const _ as *const c::sockaddr
+    }
+}
+
+fn addr2raw(addr: &SocketAddr) -> (SocketAddrCRepr, c::socklen_t) {
     match *addr {
-        SocketAddr::V4(ref a) => {
-            (a as *const _ as *const _, mem::size_of_val(a) as c::socklen_t)
+        SocketAddr::V4(addr) => {
+            // `s_addr` is stored as BE on all machine and the array is in BE order.
+            // So just transmuting the octects to the `u32` representation works.
+            #[cfg(unix)]
+            let sin_addr = c::in_addr {
+                s_addr: unsafe { mem::transmute::<_, u32>(addr.ip().octets()) },
```

Comment:
Will the alignment requirements of `u32` be fulfilled by the array emitted from `ip().octets()`?  If that isn't guaranteed, is it really that expensive to do with safe rust?

Reply:
I'm not 100% sure if this is true for all platforms, but I think it should be. The alignment of `Ipv4Addr` is 4, which is the same as `u32`. So the alignment of the `[u8; 4]` should be correct unless I'm mistaken.

We could solve it safely by using the built in `Into<u32>` etc. And the compiler might be able to optimize it away, but if not it will do a byte swap twice.

Reply:
I think avoiding the `unsafe` is worth the minuscule cost of a double byte-swap (especially considering it might be optimized away).

Reply:
Correctness is of course way more important than performance. So if we can't say for sure that this is correct I agree we should change to a safer solution.

Reply:
Fixed. Less `unsafe` is always nice!

Reply:
Oops. I realize `Ipv4Addr::octets` of course return an owned `[u8; 4]` and not a reference. As such, the IP data is copied onto the stack and the alignment is just 1, not 4. Good we changed it.